tests/msc2716_test.go

```diff
@@ -0,0 +1,858 @@
+		//
+		// Final timeline output: ( [n] = historical chunk )
+		// (oldest) A, B, [insertion, c, d, e] [insertion, f, g, h, insertion], I, J (newest)
+		//                chunk 1              chunk 0
+		t.Run("Backfilled historical events resolve with proper state in correct order", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, map[string]interface{}{
+				"preset": "public_chat",
+				"name":   "the hangout spot",
+			})
+			alice.JoinRoom(t, roomID, nil)
+
+			// Create some normal messages in the timeline. We're creating them in
+			// two batches so we can create some time in between where we are going
+			// to backfill.
+			//
+			// Create the first batch including the "live" event we are going to
+			// insert our backfilled events next to.
+			eventIDsBefore := createMessagesInRoom(t, alice, roomID, 2)
+			eventIdBefore := eventIDsBefore[len(eventIDsBefore)-1]
+			timeAfterEventBefore := time.Now()
+
+			// wait X number of ms to ensure that the timestamp changes enough for
+			// each of the messages we try to backfill later
+			numHistoricalMessages := 6
+			time.Sleep(time.Duration(numHistoricalMessages) * timeBetweenMessages)
+
+			// Create the second batch of events.
+			// This will also fill up the buffer so we have to scrollback to the
+			// inserted history later.
+			eventIDsAfter := createMessagesInRoom(t, alice, roomID, 2)
+
+			// Insert the most recent chunk of backfilled history
+			batchSendRes := batchSendHistoricalMessages(
+				t,
+				as,
+				[]string{virtualUserID},
+				roomID,
+				eventIdBefore,
+				timeAfterEventBefore.Add(timeBetweenMessages*3),
+				"",
+				3,
+				// Status
+				200,
+			)
+			batchSendResBody := client.ParseJSON(t, batchSendRes)
+			historicalEventIDs := getEventsFromBatchSendResponseBody(t, batchSendResBody)
+			nextChunkID := getNextChunkIdFromBatchSendResponseBody(t, batchSendResBody)
+
+			// Insert another older chunk of backfilled history from the same user.
+			// Make sure the meta data and joins still work on the subsequent chunk
+			batchSendRes2 := batchSendHistoricalMessages(
+				t,
+				as,
+				[]string{virtualUserID},
+				roomID,
+				eventIdBefore,
+				timeAfterEventBefore,
+				nextChunkID,
+				3,
+				// Status
+				200,
+			)
+			batchSendResBody2 := client.ParseJSON(t, batchSendRes2)
+			historicalEventIDs2 := getEventsFromBatchSendResponseBody(t, batchSendResBody2)
+
+			var expectedEventIDOrder []string
+			expectedEventIDOrder = append(expectedEventIDOrder, eventIDsBefore...)
+			expectedEventIDOrder = append(expectedEventIDOrder, historicalEventIDs2...)
+			expectedEventIDOrder = append(expectedEventIDOrder, historicalEventIDs...)
+			expectedEventIDOrder = append(expectedEventIDOrder, eventIDsAfter...)
+			// Order events from newest to oldest
+			expectedEventIDOrder = reversed(expectedEventIDOrder)
+
+			// 2 eventIDsBefore + 6 historical events + 3 insertion events + 2 eventIDsAfter
```

Comment:
Shouldn't this be `2 eventIDsBefore + 3 historical events + 3 historical events + 2 eventIDsAfter

Reply:
I could make this more clear by making it:

```
2 eventIDsBefore + [1 insertion event + 2 historical events + 1 insertion event] + [2 historical events + 1 insertion event] + 2 eventIDsAfter
                   ^ chunk1                                                        ^ chunk2
```